File: lollys-lotto-ts-sdk/src/codegen/instructions/requestWinningNumbers.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface RequestWinningNumbersAccounts {
  /** The Solana Randomness Service program. */
  randomnessService: PublicKey
  /**
   * The account that will be created on-chain to hold the randomness request.
   * Used by the off-chain oracle to pickup the request and fulfill it.
   */
  randomnessRequest: PublicKey
  /** The TokenAccount that will store the funds for the randomness request. */
  randomnessEscrow: PublicKey
  /**
   * The randomness service's state account. Responsible for storing the
   * reward escrow and the cost per random byte.
   */
  randomnessState: PublicKey
  /** The token mint to use for paying for randomness requests. */
  randomnessMint: PublicKey
  /** The account that will pay for the randomness request. */
  payer: PublicKey
  authority: PublicKey
  lottoGame: PublicKey
  eventEmitter: PublicKey
  /** The Solana System program. Used to allocate space on-chain for the randomness_request account. */
  systemProgram: PublicKey
  /** The Solana Token program. Used to transfer funds to the randomness escrow. */
  tokenProgram: PublicKey
  /** The Solana Associated Token program. Used to create the TokenAccount for the randomness escrow. */
  associatedTokenProgram: PublicKey
}

export function requestWinningNumbers(
  accounts: RequestWinningNumbersAccounts,
  programId: PublicKey = PROGRAM_ID
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.randomnessService, isSigner: false, isWritable: false },
    { pubkey: accounts.randomnessRequest, isSigner: true, isWritable: true },
    { pubkey: accounts.randomnessEscrow, isSigner: false, isWritable: true },
    { pubkey: accounts.randomnessState, isSigner: false, isWritable: false },
    { pubkey: accounts.randomnessMint, isSigner: false, isWritable: false },
    { pubkey: accounts.payer, isSigner: true, isWritable: true },
    { pubkey: accounts.authority, isSigner: false, isWritable: false },
    { pubkey: accounts.lottoGame, isSigner: false, isWritable: false },
    { pubkey: accounts.eventEmitter, isSigner: false, isWritable: true },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
    { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
    {
      pubkey: accounts.associatedTokenProgram,
      isSigner: false,
      isWritable: false,
    },
  ]
  const identifier = Buffer.from([22, 6, 33, 189, 118, 179, 35, 213])
  const data = identifier
  const ix = new TransactionInstruction({ keys, programId, data })
  return ix
}

File: programs/lollys-lotto/src/instructions/switchboard/request_winning_numbers.rs
use anchor_lang::prelude::*;
use solana_randomness_service::program::SolanaRandomnessService;
use solana_randomness_service::TransactionOptions;
use switchboard_solana::{utils::get_ixn_discriminator, NativeMint};

use anchor_spl::{
    associated_token::AssociatedToken,
    token::{Mint, Token},
};

use crate::pda_identifier::PDAIdentifier;
use crate::state::{
    EventEmitter, LollysLottoProgramEventData, LottoGame, RequestWinningNumbersEvent,
};

#[derive(Accounts)]
pub struct RequestWinningNumbers<'info> {
    /// The Solana Randomness Service program.
    pub randomness_service: Program<'info, SolanaRandomnessService>,

    /// The account that will be created on-chain to hold the randomness request.
    /// Used by the off-chain oracle to pickup the request and fulfill it.
    /// CHECK: todo
    #[account(
        mut,
        signer,
        owner = system_program.key(),
        constraint = randomness_request.data_len() == 0 && randomness_request.lamports() == 0,
    )]
    pub randomness_request: AccountInfo<'info>,

    /// The TokenAccount that will store the funds for the randomness request.
    /// CHECK: todo
    #[account(
        mut,
        owner = system_program.key(),
        constraint = randomness_escrow.data_len() == 0 && randomness_escrow.lamports() == 0,
    )]
    pub randomness_escrow: AccountInfo<'info>,

    /// The randomness service's state account. Responsible for storing the
    /// reward escrow and the cost per random byte.
    #[account(
        seeds = [b"STATE"],
        bump = randomness_state.bump,
        seeds::program = randomness_service.key(),
    )]
    pub randomness_state: Box<Account<'info, solana_randomness_service::State>>,

    /// The token mint to use for paying for randomness requests.
    #[account(address = NativeMint::ID)]
    pub randomness_mint: Account<'info, Mint>,

    /// The account that will pay for the randomness request.
    #[account(mut)]
    pub payer: Signer<'info>,

    /// CHECK: Authority of the LottoGame instance
    pub authority: AccountInfo<'info>,

    #[account(
        has_one = authority,
        seeds = [
            LottoGame::IDENT,
            authority.key().as_ref(),
            lotto_game.load()?.round.to_le_bytes().as_ref(),
        ],
        bump = lotto_game.load()?.bump,
    )]
    pub lotto_game: AccountLoader<'info, LottoGame>,

    #[account(mut)]
    pub event_emitter: Box<Account<'info, EventEmitter>>,

    /// The Solana System program. Used to allocate space on-chain for the randomness_request account.
    pub system_program: Program<'info, System>,

    /// The Solana Token program. Used to transfer funds to the randomness escrow.
    pub token_program: Program<'info, Token>,

    /// The Solana Associated Token program. Used to create the TokenAccount for the randomness escrow.
    pub associated_token_program: Program<'info, AssociatedToken>,
}

pub fn request_winning_numbers(
    ctx: Context<RequestWinningNumbers>,
) -> anchor_lang::prelude::Result<()> {
    msg!("Requesting randomness...");

    // Call the randomness service and request a new value
    solana_randomness_service::cpi::simple_randomness_v1(
        CpiContext::new(
            ctx.accounts.randomness_service.to_account_info(),
            solana_randomness_service::cpi::accounts::SimpleRandomnessV1Request {
                request: ctx.accounts.randomness_request.to_account_info(),
                escrow: ctx.accounts.randomness_escrow.to_account_info(),
                state: ctx.accounts.randomness_state.to_account_info(),
                mint: ctx.accounts.randomness_mint.to_account_info(),
                payer: ctx.accounts.payer.to_account_info(),
                system_program: ctx.accounts.system_program.to_account_info(),
                token_program: ctx.accounts.token_program.to_account_info(),
                associated_token_program: ctx.accounts.associated_token_program.to_account_info(),
            },
        ),
        6, // Request 6 bytes of randomness
        solana_randomness_service::Callback {
            program_id: crate::ID,
            accounts: vec![
                AccountMeta::new_readonly(ctx.accounts.randomness_state.key(), true).into(),
                AccountMeta::new_readonly(ctx.accounts.randomness_request.key(), false).into(),
                AccountMeta::new_readonly(ctx.accounts.authority.key(), false).into(),
                AccountMeta::new_readonly(ctx.accounts.lotto_game.key(), false).into(),
                AccountMeta::new(ctx.accounts.event_emitter.key(), false).into(),
            ],
            ix_data: get_ixn_discriminator("process_winning_numbers").to_vec(), // TODO: hardcode this discriminator [190,217,49,162,99,26,73,234]
        },
        Some(TransactionOptions {
            compute_units: Some(1_000_000),
            compute_unit_price: Some(100),
        }),
    )?;
    // Here we can emit some event to index our requests
    let event_emitter = &mut ctx.accounts.event_emitter;
    let lotto_game = &mut ctx.accounts.lotto_game;
    let block_time = Clock::get()?.unix_timestamp;
    event_emitter.emit_new_event(
        Some(block_time),
        LollysLottoProgramEventData::RequestWinningNumbers(RequestWinningNumbersEvent {
            lotto_game: lotto_game.key(),
            round: lotto_game.load()?.round,
        }),
    )?;

    Ok(())
}
